src/Markup/Avalonia.Markup.Xaml/Extensions.cs

```diff
@@ -11,38 +10,38 @@ namespace Avalonia.Markup.Xaml
 {
     internal static class Extensions
     {
-        public static T GetService<T>(this IServiceProvider sp) => (T)sp?.GetService(typeof(T));
-        
-        
-        public static Uri GetContextBaseUri(this IServiceProvider ctx) => ctx.GetService<IUriContext>().BaseUri;
+        public static T? GetService<T>(this IServiceProvider sp) => (T?)sp.GetService(typeof(T));
+
+        public static T GetRequiredService<T>(this IServiceProvider sp)
+            => sp.GetService<T>() ?? throw new InvalidOperationException($"Service {typeof(T)} hasn't been registered");
 
-        public static T GetFirstParent<T>(this IServiceProvider ctx) where T : class 
-            => ctx.GetService<IAvaloniaXamlIlParentStackProvider>().Parents.OfType<T>().FirstOrDefault();
+        public static Uri? GetContextBaseUri(this IServiceProvider ctx) => ctx.GetService<IUriContext>()?.BaseUri;
 
-        public static T GetLastParent<T>(this IServiceProvider ctx) where T : class 
-            => ctx.GetService<IAvaloniaXamlIlParentStackProvider>().Parents.OfType<T>().LastOrDefault();
+        public static T? GetFirstParent<T>(this IServiceProvider ctx) where T : class
+            => ctx.GetService<IAvaloniaXamlIlParentStackProvider>()?.Parents.OfType<T>().FirstOrDefault();
+
+        public static T? GetLastParent<T>(this IServiceProvider ctx) where T : class
+            => ctx.GetService<IAvaloniaXamlIlParentStackProvider>()?.Parents.OfType<T>().LastOrDefault();
 
         public static IEnumerable<T> GetParents<T>(this IServiceProvider sp)
-        {
-            return sp.GetService<IAvaloniaXamlIlParentStackProvider>().Parents.OfType<T>();
-        }
+            => sp.GetService<IAvaloniaXamlIlParentStackProvider>()?.Parents.OfType<T>() ?? Enumerable.Empty<T>();
 
         public static bool IsInControlTemplate(this IServiceProvider sp) => sp.GetService<IAvaloniaXamlIlControlTemplateProvider>() != null;
 
         [RequiresUnreferencedCode(TrimmingMessages.XamlTypeResolvedRequiresUnreferenceCodeMessage)]
-        public static Type ResolveType(this IServiceProvider ctx, string namespacePrefix, string type)
+        public static Type ResolveType(this IServiceProvider ctx, string? namespacePrefix, string type)
         {
             var tr = ctx.GetService<IXamlTypeResolver>();
             string name = string.IsNullOrEmpty(namespacePrefix) ? type : $"{namespacePrefix}:{type}";
-            return tr?.Resolve(name);
+            return tr?.Resolve(name) ?? throw new XamlLoadException($"Could not find type '{name}'.");
```

Comment:
Is throwing correct here? The current implementation has `tr?.Resolve` suggesting that a null is a valid result, and the usages seem to accept null.

Reply:
Thanks for taking the time to review!

Here are the usages, (hopefully I didn't miss any):
 - In `AvaloniaPropertyTypeConverter.TryResolveOwnerByName`, which also throws on a null return.
 - In `ExpressionParser` (passed as the `TypeResolver` through the chain `ReflectionBindingExtension ⇒ Binding ⇒ ExpressionObserverBuilder ⇒ ExpressionParser`):
     - `ParseTypeCastNode`: throws "Unable to determine type for cast." if the resolved type is null.
     - `ParseAttachedProperty`: the resolved type is passed to `AvaloniaPropertyRegistry.FindRegistered`, which immediately throws an `ArgumentNullException` if null.
     - `ParseFindAncestor`: the resolved type is passed to `new FindAncestorNode`, which indeed accepts null, but null here should mean that an ancestor type hasn't been specified by the user at all.
	 
Currently the only implementation of `IXamlTypeResolver` is `XamlIlRuntimeHelpers.XamlTypeResolver` which never returns null (that's why `IXamlTypeResolver.Resolve` is non-nullable). It always throws if the type can't be resolved: so we won't even get to all the extra throws.

Consequently `null` could be returned here only if there's no type resolver registered (`tr is null`).

If you agree with the comments above, I propose to simply use `GetRequiredService` instead, making what's wrong (no service registered) more obvious.

Reply:
Yep, using `GetRequiredService` makes sense to me!